### PR TITLE
Setup auth by cancancan

### PR DESCRIPTION
**Pull request by stephenkati**
- Installed `CanCanCan`.
- Added `admin` role column to the users table.
- A user can `delete` a post if it is theirs or if they have an admin role.
- A user can `delete` a comment if it is theirs or if they have an admin role.

### CODE REVIEW

File: app/models/ability.rb [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+    can :read, Post, public: true
+    can :read, User, public: true
+
+    return unless user.present?
+
+    can :manage, :all
+
+    return unless user.admin?
+
+    can :manage, :all
```
Comment (nahnahdev) [inline app/models/ability.rb]: - [ ] Great job so far @stephenkati and @0sugo There are just a few changes I will suggest you implement and you'll be on your way to the next milestone :wink: 
- [ ] Great job including `cancancan` in your project and ensuring that the `ability.rb` file was utilized :clap: However, I think that your authorization logic might not have been well put together.
- [ ] From the project requirement below, you can see that you needed to ensure that the `CanCanCan` ability class defines the set of rules for user-post interactions.
- A user can delete a post if it is theirs or if they have an admin role
- A user can delete a comment if it is theirs or if they have an admin role
- Currently, a user can delete a post or a comment even if it's not theirs. I also noticed that the `role` column is `nil` for all the created users which means none of them are `admins` and as a result shouldn't be able to delete other users' posts and comments. So kindly fix this so as to be in accordance with the project requirements :+1: 
- [ ] From the video, you can see that I was signed in as Tyra but I was able to see the delete button on the user Nahnah's Post and comment. I was also able to delete Nahnah's comment but I was unable to delete Nahnah's post and this threw an error when I tried. So kindly ensure that the delete button is not visible to a user who isn't the author of a post or comment so users aren't able to delete other users' posts and comments  :+1: 
- Video Reference

[Screencast from 07-26-2023 01:54:06 PM.webm](https://github.com/stephenkati/blog/assets/102699958/5a08ad64-0908-4967-8de8-abd5a61c5246)

- Project Requirement 
image

File: app/controllers/posts_controller.rb [hunk 2]
```diff
@@ -27,6 +27,18 @@ def create
...
+    @post = @user.posts.find_by(id: params[:id])
+
+    puts @post.inspect
+    if @post.destroy
+      redirect_to user_posts_path(@user), notice: 'Post deleted successfully!'
+    else
+      redirect_to user_posts_path(@user), alert: 'Failed to delete the post.'
+    end
+  end
+
```
Comment (nahnahdev) [inline app/controllers/posts_controller.rb:43]: - [ ] I also noticed that in your `posts_controller`, it does not look like the CanCanCan gem is being used to handle post-authorization. There is no reference to `load_and_authorize_resource` or any other CanCanCan-related methods in your code. So kindly fix this :+1:
- You can simply Load and authorize resources in the PostsController as shown below :+1: 
```ruby
class PostsController < ApplicationController
  load_and_authorize_resource

  # ... other controller methods ...
end
```

**Review by nahnahdev — CHANGES_REQUESTED** [review]
## Project Status: CHANGES REQUIRED :recycle: 
Hi @stephenkati and @0sugo , 

Good job so far!
There are some issues that you still need to work on to go to the next project but you are almost there!


## Highlights :fire: 
- Linters well setup and passing :heavy_check_mark: 
- Professional README :heavy_check_mark: 
- `cancancan` gem present in Gemfile :heavy_check_mark: 
- `role` column exist in the users table :heavy_check_mark: 
- `ability` class that defines the set of rules for user is present :heavy_check_mark: 
- `user.admin?` method created inside the user model :heavy_check_mark: 
- Destroy actions implemented in the `PostsController` and in the `CommentsController` :heavy_check_mark: 

## Required Changes ♻️
_Check the comments under the review._

## Optional suggestions
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you take them into account as they can make your code better._

You can also consider:
- N/A

Cheers, and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Review by mukhammadrizooff — APPROVED** [review]
Hi @TEAM, 

Your project is complete!  There is nothing else to say other than... it's time to merge it :shipit: 
Congratulations! 🎉

 

## Highlights
- No linter errors ✔️ 
- Correct GitFlow ✔️ 
- Good implementations ✔️
- Descriptive `README` file ✔️

## Optional suggestions

_Every comment with the [OPTIONAL] prefix won't stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better. Some of them were simply missed by the previous reviewer and addressing them will really improve your application._

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Pull request merged**